### DISCUSSION TITLE
6.3.4 - Community fixes

### DISCUSSION THREAD

**Discussion by jmcollin78** [jmcollin78/versatile_thermostat#549] [Announcements]
The release fixes some important bugs and improvements. Its installation is recommanded. There is an open discussion to follow this release if needed.

## What's Changed
* Percentage of valve opening blocked between 0 and the regulation threshold - #533 - by @jmcollin78 in https://github.com/jmcollin78/versatile_thermostat/pull/542
* try to fix "Cooling broken with Google Nest after upgrade -> 6.3.0" - #518 - by @jmcollin78 in https://github.com/jmcollin78/versatile_thermostat/pull/543
* Energy consumption value reduces during restart causing negative power - #485 - by @jmcollin78 in https://github.com/jmcollin78/versatile_thermostat/pull/544
* Service-call climate.toggle always turn A/C units to Cool when toggled ON. Regardless og previous hvac_mode when toggled OFF - and make this coherent with Windows open/close - #465 - by @jmcollin78 in https://github.com/jmcollin78/versatile_thermostat/pull/545
* vtherm doesn't follow underlying - #478 - by @jmcollin78 in https://github.com/jmcollin78/versatile_thermostat/pull/548

**Full Changelog**: https://github.com/jmcollin78/versatile_thermostat/compare/6.3.3...6.3.4

<hr /><em>This discussion was created from the release <a href='https://github.com/jmcollin78/versatile_thermostat/releases/tag/6.3.4'>6.3.4 - Community fixes</a>.</em>

**Comment by gunmalmg** [comment]
Hello

I have experienced some new issues with the min/max reg temp problems we experienced after the 2024.8 where HA Core system not accepted min/max set_temp_values outside the predefined boundaries of the climate entities temperature range. 

With the VT 6.3.4 (and 2024.10.2) I experienced the following - regarding Strong regulation of Heat pumps:

When the Strong regulation algoritme sets the target temp(the regulated_temp) on the underlaying climate entity to a value lower than the predefined minimum boundary value of the entity, It results in the following: 
The set_temp of the underlaying climate entity gets changed automatically up to the minimum set-temp value. (I assume this is done in the background by HA).
This - in terms - alter the set-temp value of the VT and my desired set-temp gets altered to the climate entities minimum boundary temp. And I loose my desired set-temp (without knowing it).

Example:
Min/max boundary on the climate entity is 16/30celcius. 
My set-temp on the VT is 19c
The room gets heated by the sun so the currant_temp is 24c. The Strong regulation reacts by lowering the regulated_temp to 15c and sets 15c as the set-temp for the underlaying climate entity. 
Since 15c i lower than the min boundary temp of the underlaying climate entity(whits is 16c), HA reacts by setting the set-temp of then underlaying climate entity to 16c. 
And so on.

Solution: 
The regulated temperature(Strong regulation for my part) must not shoot outside the min/max boundaries of the underlying climate entity. 
I think this will solve the issue. 

Regards Gunnar

**Comment by jmcollin78** [comment]
Hello @gunmalmg 

I understand your but I'm pretty sure the boundaries is respected. I will have a check more closely. You have Strong regulation if I understand well.

**Comment by gunmalmg** [comment]
Yes - I use Strong Regulation

**Comment by gunmalmg** [comment]
![E8B200CE-CE78-410B-92D1-FAB08A1E99D7](https://github.com/user-attachments/assets/7cc5645b-9738-44ca-a380-a0da33a43af9)
![42A5EB2C-DFFC-4DC3-B0C4-F92FFFDAF2B5](https://github.com/user-attachments/assets/edee39a1-0f97-4276-953d-ae26c0a57c92)
![20DB9C99-50A8-43F3-A4E8-0DC7682E4D28](https://github.com/user-attachments/assets/6057a089-31f3-478c-833e-febf9225c0a4)

**Comment by gunmalmg** [comment]
![686EF073-193D-4D56-9337-0BE36CDB4975](https://github.com/user-attachments/assets/faf07a35-f84d-4f04-a5e2-24d1c3ff6b8a)

**Comment by gunmalmg** [comment]
@jmcollin78 
I now think it has something to do with the underlying climate entity that behaves strangely. 
I look deeper into that myself- what could have possibly caused this misbehavior 
Regards Gunnar

**Comment by jmcollin78** [comment]
Hello @gunmalmg ,

looking at the code:

The method that set the temperature of the underlying:
```
async def set_temperature(self, temperature, max_temp, min_temp):
        """Set the target temperature"""
        if not self.is_initialized:
            return

        # Issue 508 we have to take care of service set_temperature or set_range
        target_temp = self.cap_sent_value(temperature)                                              <----- cap the value so that it will be in the range
        if (
            ClimateEntityFeature.TARGET_TEMPERATURE_RANGE
            in self._underlying_climate.supported_features
        ):
            data = {
                ATTR_ENTITY_ID: self._entity_id,
                "target_temp_high": target_temp,
                "target_temp_low": target_temp,
                # issue 518 - we should send also the target temperature, even in TARGET RANGE
                "temperature": target_temp,
            }
        else:
            data = {
                ATTR_ENTITY_ID: self._entity_id,
                "temperature": target_temp,
            }

        await self._hass.services.async_call(
            CLIMATE_DOMAIN,
            SERVICE_SET_TEMPERATURE,
            data,
        )

        self._last_sent_temperature = target_temp
```

and the capping is done like this:
```
@overrides
    def cap_sent_value(self, value) -> float:
        """Try to adapt the target temp value to the min_temp / max_temp found
        in the underlying entity (if any)"""

        if not self.is_initialized:
            return value

        # Gets the min_temp and max_temp
        if (
            self._underlying_climate.min_temp is not None
            and self._underlying_climate is not None
        ):
            min_val = TemperatureConverter.convert(
                self._underlying_climate.min_temp, self._underlying_climate.temperature_unit, self._hass.config.units.temperature_unit
            )
            max_val = TemperatureConverter.convert(
                self._underlying_climate.max_temp, self._underlying_climate.temperature_unit, self._hass.config.units.temperature_unit
            )

            new_value = max(min_val, min(value, max_val))
        else:
            _LOGGER.debug("%s - no min and max attributes on underlying", self)
            new_value = value

        if new_value != value:
            _LOGGER.info(
                "%s - Target temp have been updated due min, max of the underlying entity. new_value=%.0f value=%.0f min=%.0f max=%.0f",
                self,
                new_value,
                value,
                min_val,
                max_val,
            )

        return new_value
```

In clear, Vtherm is using the min and max of the underlying and not min and max of VTherm itself. I think this is confusing.

In your case min of underlying is 10 so, what you seems not abnormal.

**Comment by gunmalmg** [comment]
Hello @jmcollin78 

Very clarifying 

So the actual regulation-temp sendt to the underlaying is adjusted to be inside the boundaries of the min/max temp of the underlying. 

The confusing bit for me, was that the regulated_temperature.entity of the VTherm is not showing the actual adjusted regulated temperature that is used to regulate the underlaying set_temp. It shows the theoretical regulated temperature that VTherm calculates regardless of the underlaying temp boundaries. 

OK :-)

**Comment by gunmalmg** [comment]
Hello @jmcollin78 again 

Regarding my strange problem with the boundaries and the strange auto altering of the set-temp in VTherm :

The problem was created by a wrong min-val value of the underlaying climate device. 
I have 5 heat pumps at home, and 3 of them suffered from this strange problem. 
The 3 heat pumps have physical min_temp at 16c, but the actual min_temp value set in the underlying was now strangely altered to 10c ??
The whole problem was solved by getting the underlying Min_temp value set back to 16c. (that I had to achieve by altering the heat pump type in the underlaying heat pump integration.)

Sorry for the bothering of you. But educational - and somewhat bothering - for me 😜

Regards Gunnar

**Reply by jmcollin78** [reply]
No problem. Have a nice day !
